doc/pserver.md

```diff
@@ -0,0 +1,257 @@
+# Parameter Server Design
+
+
+## Overview
+
+
+## PServer
+
+
+### Compoments
+
+
+![pserver](./images/pserver.png)
+
+
+PServer contains three main compoments:
+
+- KVStore
+- Gradient Queue
+- Optimizer
+
+1. The worker initializes a model, and pushes parameters to KVStore.
+2. Before each step of training, the worker pulls the latest model from KVStore.
+3. After a round of forward/backward computation, the worker pushes gradients to the Gradient Queue waiting for processing.
+4. Optimizer gets a gradient from the Gradient Queue.
+5. Then, Optimizer looks up the corresponding parameter from KVStore.
+6. Optimizer applies gradients to parameters, and updates parameter back to KVStore.
+
+
+### Tensor Data Structure
+
+To support data communication between pods, we introduce a `Tensor` proto message:
+
+```proto
+message Tensor {
+    enum DataType {
+        BOOL = 0;
+        INT16 = 1;
+        INT32 = 2;
+        INT64 = 3;
+        FP16 = 4;
+        FP32 = 5;
+        FP64 = 6;
+    }
+    string name = 1;
+    DataType data_type = 2;
+    repeated int64 dims = 3;
+    bytes content = 4;
+    repeated int64 indices = 5;
+}
+```
+
+Correspondingly, we have a `Tensor` Python class.
+
+
+```python
+class Tensor(object):
+    def __init__(self, name=None, value=None, indices=None):
+        self.name = name
+        self.value = value
+        self.indices = indices
+```
+
+There are also some helper functions:
+
+```python
+def serialize_to_pb(tensor, pb):
+    pass
+
+def deserialize_from_pb(pb, tensor):
+    pass
+
+def convert_to_tf_variable(tensor):
+    pass
+
+def convert_to_tf_tensor(tensor):
+    pass
+```
+
+
+
+### KVStore
+
+For a common model variable, we use save it as a `tf.Variable` in Parameter DB. For the embedding table, we introduce a customized data structure.
+
+
+```python
+class EmbeddingTable(object):
```

Comment:
Here, I think that we should have a data structure of EmbeddingTable. One embedding layer has one embedding table. The embedding table now store <id, embedding_vector> pairs. This is a Python class now, we could re-implement it as a C++ structure in the future for better performance if necessary.


If there is a embedding gradient, the optimization logic could be:

```python
y, idx = tf.unique(grad.indices)
grad.indices = idx.numpy()
param = self.kvstore.get_embedding_vector(grad.name, y.numpy())
param_var = convert_to_var(param)
grads_and_params = list(zip([convert_to_tensor(grad)], [param_var]))
self.opt.apply_gradients(grads_and_params)
param.value = param_var.numpy()
self.kvstore.set_embeddint_vector(param.name, param.indices, param.value)
```

Reply:
If we only store <id, embedding_vector> pairs, we need to create multiple embedding tables for one embedding layer. Because we also have slot variables such as momentum. For slot variables, the key in redis now is `<layer_name>-<slot_name>-<id>`.

Reply:
Also, the optimization logic has implemented in `OptimizerWrapper` now.

Reply:
Yes, exactly. If we store embedding table param as an EmbeddingTable object, we also need to make embedding table mementum slot as a EmbeddingTable object.

I write a simple test script to record the time:


```python
import numpy as np
from random import randrange
from time import process_time


id_num = 100000
embedding_dim = 128
id_limit = 10000000
layer_name = "tom"

embedding_vectors = np.random.uniform(-1, 1, size=(id_num, embedding_dim))
set_ids = [randrange(id_limit) for i in range(id_num)]
get_ids = [randrange(id_limit) for i in range(id_num)]

class EmbeddingTable(object):
    def __init__(self, name):
        self.name = name
        self.vectors = {}

    def set(self, ids, value):
        for index, id in enumerate(ids):
            row = value[index,:]
            self.vectors[id] = row

    def get(self, ids):
        unknow_ids = []
        know_ids = []
        res = []
        for id in ids:
            if id not in self.vectors:
                unknow_ids.append(id)
            else:
                know_ids.append(id)
                res.append(self.vectors[id])

        res = np.stack(res)
        return res, unknow_ids



class EmbeddingTable2(object):
    def __init__(self):
        self.db = {}

    def db_set(self, ids, value):
        for index, id in enumerate(ids):
            key = layer_name + "_" + str(id)
            row = value[index,:]
            self.db[key] = row

    def db_get(self, ids):
        unknow_ids = []
        know_ids = []
        res = []
        for id in ids:
            key = layer_name + "_" + str(id)
            if key not in self.db:
                unknow_ids.append(id)            
            else:
                know_ids.append(id)
                res.append(self.db[key])

        res = np.stack(res)
        return res, unknow_ids


start = process_time()

t1 = EmbeddingTable(layer_name)
t1.set(set_ids, embedding_vectors)
t1.get(get_ids)

end = process_time()
print("%.2gs" % (end-start))


start = process_time()
t2 = EmbeddingTable2()
t2.db_set(set_ids, embedding_vectors)
t2.db_get(get_ids)

end = process_time()
print("%.2gs" % (end-start))
```


The result is 
```
0.072s
0.15s
```

The first one is almost twice faster than the second one.

Anyway, I mean that we should extract EmbeddingTable as an independent data structure. It provides`get/set` interface.  We could find better and better concrete implementations. Or even we will implement it in C++, and expose it to Python later.